src/brevitas/nn/quant_convtranspose.py

```diff
@@ -218,3 +220,102 @@ def max_acc_bit_width(self, input_bit_width, weight_bit_width):
+            self,
+            in_channels=in_channels,
+            out_channels=out_channels,
+            kernel_size=kernel_size,
+            stride=stride,
+            padding=padding,
+            output_padding=output_padding,
+            dilation=dilation,
+            groups=groups,
+            padding_mode=padding_mode,
+            bias=bias,
+            device=device,
+            dtype=dtype)
+        QuantWBIOL.__init__(
+            self,
+            weight_quant=weight_quant,
+            bias_quant=bias_quant,
+            input_quant=input_quant,
+            output_quant=output_quant,
+            return_quant_tensor=return_quant_tensor,
+            **kwargs)
+        self._output_size = None
+
+    @property
+    def per_elem_ops(self):
+        raise NotImplementedError
+
+    @property
+    def output_channel_dim(self) -> int:
+        return 1
+
+    @property
+    def channelwise_separable(self) -> bool:
+        raise self.groups == self.out_channels
+
+    def forward(self,
+                input: Union[Tensor, QuantTensor],
+                output_size=None) -> Union[Tensor, QuantTensor]:
+        self._output_size = output_size  # cache the value temporarily
+        return self.forward_impl(input)
+
+    def compute_output_padding(self, inp, output_size):
+        if torch_version >= version.parse('1.12'):
+            return self._output_padding(
+                inp, output_size, self.stride, self.padding, self.kernel_size, num_spatial_dims=3)
+        else:
+            return self._output_padding(
+                inp, output_size, self.stride, self.padding, self.kernel_size)
+
+    def conv_transpose3d_zeros_pad(
+            self, x: Tensor, weight: Tensor, bias: Optional[Tensor], output_padding):
+        out = conv_transpose3d(
+            x, weight, bias, self.stride, self.padding, output_padding, self.groups, self.dilation)
+        return out
+
+    def inner_forward_impl(self, x: Tensor, quant_weight: Tensor, quant_bias: Optional[Tensor]):
+        if self.padding_mode == 'zeros':
+            output_padding = self.compute_output_padding(x, self._output_size)
+            self._output_size = None  # set it back to None after consuming it
+            out = self.conv_transpose3d_zeros_pad(x, quant_weight, quant_bias, output_padding)
+            return out
+        else:
+            raise NotImplementedError(f"Padding mode {self.padding_mode} not supported.")
+
+    def max_acc_bit_width(self, input_bit_width, weight_bit_width):
+        max_uint_input = max_int(bit_width=input_bit_width, signed=False, narrow_range=False)
+        max_kernel_val = self.weight_quant.max_uint_value(weight_bit_width)
+        group_size = self.out_channels // self.groups
+        overlapping_sums = max(round(self.kernel_size[0] / self.stride[0]), 1)
+        overlapping_sums *= max(round(self.kernel_size[1] / self.stride[1]), 1)
+        overlapping_sums *= max(round(self.kernel_size[2] / self.stride[2]), 1)
```

Comment:
Does this actually calculate the right thing? The way I read this function is that `group_size` * `overlapping_sums` should equal the number of terms in my dot product. If so, why is the stride important?

@i-colbert, @Giuseppe5, can you enlighten me here?

Reply:
I have only worked with the 2D case, but if it is a trivial extension of that then this is almost correct. Unlike the convolution, which traverses the input space in strides, the deconvolution traverses the output space in strides and creates overlapping sums that are a function of the kernel size and stride.  [This paper](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=9592768) introduces an algorithm the traverses the output space of the deconvolution (see Algorithm 8) and you can see the stride skipping there. However, you can see that it should be a floor rounding rather than a halfway rounding. I would propose the following:

```python
...
patch_size = (self.kernel_size[0] // self.stride[0]) * (self.kernel_size[1] // self.stride[1]) * (self.kernel_size[2] // self.stride[2])
max_uint_output = max_uint_input * max_kernel_val * patch_size * group_size
...
```

The kernel size shouldn't really outgrow the stride, since it wouldn't make much sense unless there is a niche use case. Not sure PyTorch throws an error in that case and also not sure if it'd be the job of the quant layer to protect against this.

Reply:
Thank you! <3